src/ray/core_worker/task_manager.h

```diff
@@ -50,23 +55,33 @@ class TaskManager : public TaskFinisherInterface {
   void CompletePendingTask(const TaskID &task_id,
                            const rpc::PushTaskReply &reply) override;
 
-  /// Treat a pending task as failed.
+  /// A pending task failed. This will either retry the task or mark the task
+  /// as failed if there are no retries left.
   ///
   /// \param[in] task_id ID of the pending task.
   /// \param[in] error_type The type of the specific error.
-  /// \return Void.
-  void FailPendingTask(const TaskID &task_id, rpc::ErrorType error_type) override;
+  void PendingTaskFailed(const TaskID &task_id, rpc::ErrorType error_type) override;
 
  private:
+  /// Treat a pending task as failed. The lock should not be held when calling
+  /// this method because it may trigger callbacks in this or other classes.
+  void MarkPendingTaskFailed(const TaskID &task_id, int64_t num_returns,
+                             rpc::ErrorType error_type) LOCKS_EXCLUDED(mu_);
+
   /// Used to store task results.
   std::shared_ptr<CoreWorkerMemoryStore> in_memory_store_;
 
+  /// Called when a task should be retried.
+  const RetryTaskCallback retry_task_callback_;
+
   /// Protects below fields.
   absl::Mutex mu_;
 
-  /// Map from task ID to the task's number of return values. This map contains
-  /// one entry per pending task that we submitted.
-  absl::flat_hash_map<TaskID, int64_t> pending_tasks_ GUARDED_BY(mu_);
+  /// Map from task ID to a pair of:
+  ///   {task spec, number of allowed retries left}
+  /// This map contains one entry per pending task that we submitted.
+  absl::flat_hash_map<TaskID, std::pair<TaskSpecification, int>> pending_tasks_
```

Comment:
```suggestion
  absl::flat_hash_map<TaskID, std::pair<std::shared_ptr<TaskSpecification>, int>> pending_tasks_
```

Please make TaskSpec as shared_ptr as copy it is very expensive.

Reply:
TaskSpec already contains a shared_ptr to the protobuf, so it is meaningless to have another shared_ptr to TaskSpec on top of that.